crates/sparrow-instructions/src/evaluators/field_ref.rs

```diff
@@ -30,38 +64,51 @@ impl EvaluatorFactory for FieldRefEvaluator {
                 );
             };
 
-        let base_fields = if let DataType::Struct(fields) = input_type {
-            fields
-        } else {
-            anyhow::bail!(
-                "Unable to create FieldRefEvaluator for input type {:?}",
-                &info.args[0].data_type
-            );
-        };
-
-        let field_index = base_fields
-            .iter()
-            .position(|field| field.name() == field_name)
-            .ok_or_else(|| anyhow!("No field named '{}' in struct {:?}", field_name, input_type))?;
-
+        let (field_index, field) = get_field_index(field_name, input_type)?;
         let (base, _) = info.unpack_arguments()?;
-        Ok(Box::new(Self { base, field_index }))
+        Ok(Box::new(Self {
+            base,
+            field_index,
+            field,
+        }))
     }
 }
 
 impl Evaluator for FieldRefEvaluator {
     fn evaluate(&mut self, info: &dyn crate::RuntimeInfo) -> anyhow::Result<ArrayRef> {
-        let input = info.value(&self.base)?.struct_array()?;
-
-        let field_array = input.column(self.field_index);
-
-        // Arrow field ref ignores the null-ness of the outer struct.
-        // To handle this, we null the field out if the struct was null.
-        // TODO: Express in the DFG so we can avoid computing `is_not_null` repeatedly?
-        let is_struct_null = arrow::compute::kernels::boolean::is_null(input.as_ref())?;
-        // We need to null out rows that are null in the struct, since
-        // they may not be properly nulled-out in the field column.
-        arrow::compute::nullif(field_array.as_ref(), &is_struct_null)
-            .context("null_if for field ref")
+        let input = info.value(&self.base)?.array_ref()?;
+
+        match input.data_type() {
+            DataType::Struct(_) => {
+                let input = input.as_struct();
+                let field = input.column(self.field_index);
+
+                // Arrow field ref ignores the null-ness of the outer struct.
+                // To handle this, we null the field out if the struct was null.
+                let nulls = NullBuffer::union(input.nulls(), field.nulls());
+
+                let data = field.to_data().into_builder().nulls(nulls).build()?;
+                Ok(make_array(data))
+            }
+            DataType::List(_) => {
```

Comment:
The logic for struct and list are similar -- the only difference is creating the new list array using the same offsets. We would do the same for Map (using the same key).